# Promise

- 동기 작업의 최종 완료 또는 실패를 나타내는 Array나 Object 처럼 독자적인 객체
- 비동기 작업이 끝날 때까지 결과를 기다리는 것이 아니라, 결과를 제공하겠다는 '약속'을 반환한다는 의미에서 Promise라 명명

## promise 객체 사용법

### 프로미스 객체 생성

Promise 객체를 생성하려면 new 키워드와 Promise 생성자 함수를 사용하면 됨
이때 Promise 생성자 안에 두개의 매개변수를 가진 콜백 함수를 넣게 되는데

→ 첫 번째 인수는 작업이 성공했을 때 성공(resolve)임을 알려주는 객체이며, 두 번째 인수는 작업이 실패했을 때 실패(reject)임을 알려주는 오류 객체임

<aside>
💡

Promise 생성자안에 들어가는 콜백 함수를 executor 라고 부름

</aside>

```jsx
const myPromise = new Promise((resolve, reject) => {
  // 비동기 작업 수행
  const data = fetch("서버로부터 요청할 URL");

  if (data)
    resolve(data); // 만일 요청이 성공하여 데이터가 있다면
  else reject("Error"); // 만일 요청이 실패하여 데이터가 없다면
});
```

### 프로미스 객체 처리

이렇게 만들어진 Promise 객체는 비동기 작업이 완료된 이후에 다음 작업을 연결시켜 진행할 수 있다.

작업 결과 따라 .then() 과.catch() 메서드 체이닝을 통해 성공과 실패에 대한 후속 처리를 진행할 수 있다.

만일 처리가 성공하여 프로미스 객체 내부에서resolve(data) 를 호출하게 되면, 바로.then() 으로 이어져then 메서드의 콜백 함수에서 성공에 대한 추가 처리를 진행한다.

이때 호출한resolve() 함수의 매개변수의 값이then 메서드의 콜백 함수 인자로 들어가then 메서드 내부에서 프로미스 객체 내부에서 다룬 값을 사용할 수 있게 된다.

반대로 처리가 실패하여 프로미스 객체 내부에서 reject("Error") 를 호출하게 되면, 바로.catch() 로 이어져catch 메서드의 콜백 함수에서 성공에 대한 추가 처리를 진행한다.

```jsx
myPromise
    .then((value) => { // 성공적으로 수행했을 때 실행될 코드
    	console.log("Data: ", value); // 위에서 return resolve(data)의 data값이 출력된다
    })
    .catch((error) => { // 실패했을 때 실행될 코드
     	console.error(error); // 위에서 return reject("Error")의 "Error"가 출력된다
    })
    .finally(() => { // 성공하든 실패하든 무조건 실행될 코드

```
